File: src/main/java/org/taktik/icure/dao/replicator/GroupDBUrl.java
package org.taktik.icure.dao.replicator;

import org.taktik.icure.entities.Group;

import java.net.URI;
import java.net.URISyntaxException;

/**
 * @author Bernard Paulus - 14/03/2017
 */
public class GroupDBUrl {
    private final URI couchDbUrl;

    public GroupDBUrl(String couchDbUrl) {
        try {
            this.couchDbUrl = new URI(couchDbUrl);
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("malformed couchdb url: " + couchDbUrl, e);
        }
    }

    public String getDbName(Group group) {
        return "icure-" + group.getId() + "-base";
    }

    public String getInstanceUrl(Group group) {
        URI result;
        try {
            result = new URI(couchDbUrl.getScheme(),
                    null,
                    couchDbUrl.getHost(),
                    couchDbUrl.getPort(),
                    null,
                    null,
                    null);
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("failed to build url", e);
        }
        return result.toString();
    }

    public String getDbUrl(Group group) {
        URI result;
        try {
            result = new URI(couchDbUrl.getScheme(),
                    group.getId() + ":" + group.getPassword(),
                    couchDbUrl.getHost(),
                    couchDbUrl.getPort(),
                    "/" + getDbName(group),
                    null,
                    null);
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("failed to build url", e);
        }
        return result.toString();
    }

    public String getLocalDbUrl(Group group) {
        URI result;
        try {
            result = new URI(couchDbUrl.getScheme(),
                    group.getId() + ":" + group.getPassword(),
                    "127.0.0.1",
                    couchDbUrl.getPort(),
                    "/" + getDbName(group),
                    null,
                    null);
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("failed to build url", e);
        }
        return result.toString();
    }


}

File: src/main/kotlin/org/taktik/icure/dao/replicator/GroupDBUrl.kt
package org.taktik.icure.dao.replicator

import org.taktik.icure.entities.Group

import java.net.URI
import java.net.URISyntaxException
import java.net.URL

class GroupDBUrl(couchDbUrl: String) {
    private val couchDbUrl = URI(couchDbUrl)

    fun getDbName(group: Group): String {
        return "icure-" + group.id + "-base"
    }

    fun getInstanceUrl(group: Group): String =
        try {
            val server = group.servers?.firstOrNull()?.let { URI(it) } ?: couchDbUrl

            URI(server.scheme,
                null,
                    server.host,
                    server.port, null, null, null).toString()
        } catch (e: URISyntaxException) {
            throw IllegalArgumentException("failed to build url", e)
        }

}
